gprofiler/profilers/java.py

```diff
@@ -257,14 +257,45 @@ def is_java_basename(process: Process) -> bool:
 _JAVA_VERSION_CACHE_MAX = 1024
 
 
+def _get_process_ns_java_path(process: Process) -> Optional[str]:
+    """
+    Look up path to java executable installed together with this process' libjvm.
+    """
+    java_path: Optional[str] = None
+    nspid = get_process_nspid(process.pid)
+    if is_java_basename(process):
+        java_path = f"/proc/{nspid}/exe"  # it's a symlink and will be resolveable under process' mnt ns
+    else:
+        libjvm_path: Optional[str] = None
+        for m in process.memory_maps():
+            if re.match(DETECTED_JAVA_PROCESSES_REGEX, m.path):
+                libjvm_path = m.path
+                break
+        if libjvm_path is None:
+            return None
+        libjvm_dir = os.path.dirname(libjvm_path)
+        # support two java layouts - it's either lib/server/../../bin/java or lib/{arch}/server/../../../bin/java:
+        java_candidate_paths = [
+            Path(libjvm_dir, "../../bin/java").resolve(),
+            Path(libjvm_dir, "../../../bin/java").resolve(),
+        ]
+        for p in java_candidate_paths:
+            proc_relative_path = Path(f"/proc/{process.pid}/root", p.relative_to("/"))
+            if proc_relative_path.exists():
+                if os.access(proc_relative_path, os.X_OK):
+                    java_path = os.path.join(f"/proc/{nspid}/root", p)
+                    break
+    return java_path
+
+
 # process is hashable and the same process instance compares equal
 @functools.lru_cache(maxsize=_JAVA_VERSION_CACHE_MAX)
-def get_java_version(process: Process, stop_event: Event) -> str:
-    # make sure we're only called for "java" processes, otherwise running "-version" makes no sense.
-    # our callers should check for it.
-    assert is_java_basename(process), f"expected java, found {process!r}"
-
-    nspid = get_process_nspid(process.pid)
+def get_java_version(process: Process, stop_event: Event) -> Optional[str]:
+    # make sure we're able to find "java" binary bundled with process libjvm
+    process_java_path = _get_process_ns_java_path(process)
+    if process_java_path is None:
+        return None
+    java_path = process_java_path
 
     # this has the benefit of working even if the Java binary was replaced, e.g due to an upgrade.
```

Comment:
This comment is no longer relevant here - it refers to the assignment of `/proc/nspid/exe` which now happens in `_get_process_ns_java_path`. Move it to the relevant site?